optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java

```diff
@@ -247,6 +252,62 @@ protected void setCalculatedScore(Score score) {
         calculationCount++;
     }
 
+    @Override
+    public String explainScore() {
+        final int INDICTMENT_LIMIT = 5;
+        final int CONSTRAINT_MATCH_LIMIT = 2;
+        Score workingScore = calculateScore();
+        Collection<ConstraintMatchTotal> constraintMatchTotals = getConstraintMatchTotals();
+        ConstraintMatchScoreComparator constraintMatchScoreComparator = new ConstraintMatchScoreComparator();
+        StringBuilder scoreExplanation = new StringBuilder((constraintMatchTotals.size() + 4 + 2 * INDICTMENT_LIMIT) * 80);
+        scoreExplanation.append("Explanation of score (").append(workingScore).append("):\n");
+        scoreExplanation.append("    Constraint match totals:\n");
+        constraintMatchTotals.stream()
+                .sorted(new ConstraintMatchTotalScoreComparator())
+                .forEach(constraintMatchTotal -> {
```

Comment:
Explaining entire score and explaining a single ConstraintMatch are two different level of abstraction; maybe the lambda could go into a private method.